### PR TITLE
Fix chrome storage access scope

### LINKED ISSUE #9949: `chrome.storage` in injected_scripts does not share data across domains
* Electron version: 1.7.2
* Operating system: macOS (but concers all)

### Actual behavior
```js
// injected in https://mail.google.com/ for extenions ABC
chrome.storage.sync.set({foo: 'bar'})

// injected in https://mailtracker.hunter.io for extensions ABC
chrome.storage.sync.get('foo', obj => console.log(obj)) // nothing :(

// injected in https://mail.google.com/ for extension XYZ
chrome.storage.sync.get('foo', obj => console.log(obj)) // {foo: 'bar'}
```

### Expected behavior
From my [understanding of `chrome.storage`](https://developer.chrome.com/extensions/storage) the data should be scoped by extension and not by domain in which the extension's `content_scripts` are injected.

```js
// injected in https://mail.google.com/ for extenion ABC
chrome.storage.sync.set({foo: 'bar'})

// injected in https://mailtracker.hunter.io for extension ABC
chrome.storage.sync.get('foo', obj => console.log(obj)) // {foo: 'bar'}

// injected in https://mail.google.com/ for extension XYZ
chrome.storage.sync.get('foo', obj => console.log(obj)) // nothing
```

### How to reproduce

_Can provide a repo_

### Reason

Because our implementation of `chrome.storage` [uses](https://github.com/electron/electron/blob/master/lib/renderer/extensions/storage.js#L2) `window.localStorage`, the access to data is scoped to host web-page's domain.

### Suggested solution
- replace the use of `window.localStorage` in [`storage.js`](https://github.com/electron/electron/blob/master/lib/renderer/extensions/storage.js) by a file based storage (`electron-json-storage` ?)
- create and inject a store by `extensionId` same as [`extensions/i18.js`](https://github.com/electron/electron/blob/master/lib/renderer/extensions/i18n.js#L78-L84)

_Related: #1498_

### PR DESCRIPTION
**Pull request by alexstrat**
Fixes #9949

### CODE REVIEW

**Comment by MarshallOfSound** [comment]
I'm not sold on this approach personally, there are a lot of edge cases and things to handle when persisting things to disk.  You definitely want to at least be using the async fs methods with some form of back-off and retry logic for when things don't work the first time.

You also need to recursively make the parent directories if they don't exist.  

I'd much rather find a way to pass the persistence of this data off to Chromium logic that has already been battle tested

**Comment by alexstrat** [comment]
Definitely ok for using async `fs` methods and recursively `mkdir` the parent directories => will do so. That's the way to do it properly!

> I'd much rather find a way to pass the persistence of this data off to Chromium logic that has already been battle tested

Do you have somthing in sight? I think every DOM storage methods will be scoped by domain and I'm not aware of other Chromium-based persistence methods already implemented and battle-tested.

**Comment by MarshallOfSound** [comment]
> Do you have somthing in sight?

Never really looked into the implementation of storage in Chromium too much, @zcbenz might know if their is a Chromium API we can expose and use here.

**Comment by alexstrat** [comment]
I've implemented the use of async fs as well as recursive mkdir.

> ..with some form of back-off and retry logic for when things don't work the first time

I'd be glad to do it, but how and why things would not work for the first time? Other parts of electron use persisted data on disk and it does not look like to me they implement some sort of back-off. Is that because we are on renderer side? If so, wouldn't be wise to centralise these access in the main process and deal via IPCs.

---
I've looked into Chromium's storage related to `chrome.storage` and here is my interpretations:
- API exposed in renderer via [`extensions/renderer/storage_area.cc`](https://cs.chromium.org/chromium/src/extensions/renderer/storage_area.cc?sq=package:chromium&dr=C) that calls [`APIRequestHandler#StartRequest`](https://cs.chromium.org/chromium/src/extensions/renderer/bindings/api_request_handler.h?sq=package:chromium&dr=C) (A wrapper around a map for extension API calls)
- the request are (a priori) received in `extensions/browser/api/storage` (see [`storage_api.h`](https://cs.chromium.org/chromium/src/extensions/browser/api/storage/storage_api.h?sq=package:chromium&dr=C) and [`storage_api.c`](https://cs.chromium.org/chromium/src/extensions/browser/api/storage/storage_api.cc?sq=package:chromium&dr=C)
- even if I haven't found very precise proof, it lloks like the underneath file storage is handled by [`extensions/browser/value_store`](https://cs.chromium.org/chromium/src/extensions/browser/value_store) which, in the end looks like being backed by leveldb

File: lib/renderer/extensions/storage.js [hunk 1]
```diff
@@ -1,88 +1,126 @@
...
-  const json = JSON.stringify(storage)
-  window.localStorage.setItem(`__chrome.storage.${storageType}__`, json)
+// recursively create directory and parent directories if needed
+const mkdirParent = (dirPath, mode, callback) => {
+  fs.mkdir(dirPath, mode, error => {
+    if (error && error.errno === 34) {
+      fs.mkdirParent(path.dirname(dirPath), mode, callback)
+      fs.mkdirParent(dirPath, mode, callback)
+    }
+    callback && callback(error)
```
Comment (zcbenz) [inline lib/renderer/extensions/storage.js]: On the worst case `callback` might be called 3 times for one `mkdirParent` call, for async API you have to wait until the last one finishes before continue.

```js

const mkdirp = (dir, callback) => {
  fs.mkdir(dir, (error) => {
    if (error && error.code === 'ENOENT') {
      mkdirp(path.dirname(dir), (error) => {
        if (!error) {
          mkdirp(dir, callback)
        }
      })
    } else if (error && error.code === 'EEXIST') {
      callback(null)
    } else {
      callback(error)
    }
  })
}
```

**Comment by zcbenz** [comment]
This change would make old apps lose their data in storage, but I guess it is fine since it only affects devtools extensions.

File: lib/renderer/extensions/storage.js (hunk 1 above)
Comment (alexstrat) [inline lib/renderer/extensions/storage.js]: Corrected. Used your suggestion as it.

**Pull request merged**